pynamodb/attributes.py

```diff
@@ -820,18 +820,15 @@ def deserialize(self, values):
         Decode as a dict.
         """
         deserialized_dict: Dict[str, Any] = dict()
-        for k in values:
-            v = values[k]
-            attr_value = _get_value_for_deserialize(v)
-            key = self._dynamo_to_python_attr(k)
-            attr_class = self._get_deserialize_class(key, v)
-            if key is None or attr_class is None:
-                continue
-            deserialized_value = None
-            if attr_value is not None:
-                deserialized_value = attr_class.deserialize(attr_value)
-
-            deserialized_dict[key] = deserialized_value
+        for k, v in values.items():
+            attr_name = self._dynamo_to_python_attr(k)
+            attr_type, attr_value = next(iter(v.items()))
+            attr_class = self._get_deserialize_class(attr_name, attr_type)
+            if attr_class and attr_class.attr_type != attr_type:
+                raise ValueError("Cannot deserialize '{}' attribute from type: {}".format(
+                    attr_name, attr_type))
```

Comment:
Adds a check to ensure we are deserializing the correct type (previously we would send any value to `deserialize` and wait for a type error).

Reply:
👏  finally, saw it bit folks more than once

Reply:
I've seen people write custom attributes that handle multiple types to support migrations. IIRC that was not within a MapAttribute, but is it something we should still support?

Reply:
@garrettheel

Was it something like this?
```
foo_old = NumberAttribute(attr_name='foo')
foo_new =UnicodeAttribute(attr_name='foo') 
```

Reply:
Nope, more like:

```python
class NumberUnicodeAttribute(UnicodeAttribute):
    def deserialize(self, value: str) -> int:
        try:
            deserialized_value = json.loads(value)
        except json.decoder.JSONDecodeError:
            deserialized_value = value
        return int(deserialized_value)
```

Reply:
I suppose the issue there is that the attr_type being used to deserialize did not match the one stored in DDB, but it happened to work anyway. This will uncover those issues immediately by failing to deserialize, but might cause a little bit of pain.

Reply:
So supposedly there's just one `Attribute.attr_type` so an attribute has to choose whether it's an `S` or an `N`. However, you can override `Attribute.get_value` to modify this behavior **for when your attribute is directly on the the Model**. This doesn't apply nested MapAttribute deserializations, since it's not implemented symmetrically (but maybe it should be).

Reply:
I like the idea of standardizing on using `get_value` to special case this behavior and raising an exception if this returns `None`.